File: javascript/src/providers/_shared.ts
import pRetry from "p-retry";
import { z } from "zod";

import type { BillingEvent, BillingEventResult } from "../billing/index.js";
import type { BillingEventSink } from "../bursar.js";
import { BursarError, StoreUnavailableError } from "../errors.js";
import type { ExternalValue } from "../shared/json.js";

class BillingClaimBusyError extends Error {
  override readonly name = "BillingClaimBusyError";
}

const ACKNOWLEDGED_BILLING_EVENT_ERRORS = new Set([
  "unhandled_event_type",
  "account_not_found",
  "invalid_request",
  "idempotency_conflict",
  "max_retries_exceeded",
]);

type ProviderValue = ExternalValue;

export function requireProviderString(value: ProviderValue, field: string): string {
  const parsed = z.string().safeParse(value);
  if (!parsed.success || parsed.data.trim().length === 0) {
    throw new TypeError(`${field} must be a non-empty string`);
  }
  return parsed.data.trim();
}

export function optionalProviderString(value: ProviderValue, field: string): string | undefined {
  if (value === null || value === undefined) return undefined;
  return requireProviderString(value, field);
}

export function optionalProviderBoolean(value: ProviderValue, field: string): boolean | undefined {
  if (value === null || value === undefined) return undefined;
  const parsed = z.boolean().safeParse(value);
  if (!parsed.success) {
    throw new TypeError(`${field} must be a boolean`);
  }
  return parsed.data;
}

export function requireMinorUnits(value: ProviderValue, field: string, positive = false): number {
  const parsed = z.union([z.number(), z.string().regex(/^\d+$/u)]).safeParse(value);
  const amount = parsed.success ? Number(parsed.data) : Number.NaN;
  if (!Number.isSafeInteger(amount) || amount < (positive ? 1 : 0)) {
    throw new TypeError(
      `${field} must be a ${positive ? "positive" : "non-negative"} safe integer`,
    );
  }
  return amount;
}

export function requireCurrency(value: ProviderValue, field: string): string {
  const parsed = z
    .string()
    .regex(/^[A-Za-z]{3}$/u)
    .safeParse(value);
  if (!parsed.success) {
    throw new TypeError(`${field} must be a three-letter currency code`);
  }
  return parsed.data.toUpperCase();
}

/**
 * Wrapper around the facade event sink that throws on retryable or unexpected
 * failures. Permanent claim outcomes are returned so the provider can
 * acknowledge them without scheduling another delivery.
 */
export async function callBillingEventSink(
  sink: BillingEventSink,
  event: BillingEvent,
): Promise<BillingEventResult> {
  let result: BillingEventResult;
  try {
    result = await pRetry(
      async () => {
        const attempted = await sink.ingestBillingEvent(event);
        if (attempted.error === "claim_busy") throw new BillingClaimBusyError();
        return attempted;
      },
      {
        retries: 6,
        minTimeout: 25,
        maxTimeout: 800,
        factor: 2,
        randomize: true,
        maxRetryTime: 5_000,
        shouldRetry: ({ error }) => error instanceof BillingClaimBusyError,
      },
    );
  } catch (cause) {
    if (cause instanceof BillingClaimBusyError) {
      throw new StoreUnavailableError("Billing event claim remained busy", {
        cause,
        details: { reason: "claim_busy" },
      });
    }
    throw cause;
  }

  if (result.error === "claim_failed_retry") {
    throw new StoreUnavailableError("Billing event claim could not be acquired", {
      details: { reason: result.error },
    });
  }
  if (!result.handled && !ACKNOWLEDGED_BILLING_EVENT_ERRORS.has(result.error ?? "")) {
    throw new BursarError("Bursar failed to ingest the billing event", {
      details: { reason: result.error ?? "unknown" },
    });
  }
  return result;
}

File: python/src/bursar/providers/_shared.py
from __future__ import annotations

from typing import TYPE_CHECKING

from bursar.billing.types import BillingEvent, BillingEventResult, BillingSubscriptionStatus
from bursar.errors import BursarError, StoreUnavailableError
from bursar.retry import BursarRetryOptions, retry_bursar_operation

if TYPE_CHECKING:
    from bursar.billing.contracts import BillingEventSink


class _BillingClaimBusyError(Exception):
    pass


def require_provider_string(value: object, field: str) -> str:
    """Return a non-empty provider value without manufacturing an identifier."""

    if not isinstance(value, str) or not value.strip():
        raise ValueError(f"{field} must be a non-empty string")
    return value.strip()


def optional_provider_string(value: object, field: str) -> str | None:
    """Validate an optional provider string without manufacturing a value."""

    if value is None:
        return None
    return require_provider_string(value, field)


def require_minor_units(value: object, field: str, *, positive: bool = False) -> int:
    """Validate an integral provider amount before it crosses the billing boundary."""

    if isinstance(value, bool):
        raise ValueError(f"{field} must be an integer")
    if isinstance(value, int):
        amount = value
    elif isinstance(value, str) and value.isascii() and value.isdigit():
        amount = int(value)
    else:
        raise ValueError(f"{field} must be an integer")
    if amount < (1 if positive else 0):
        qualifier = "positive" if positive else "non-negative"
        raise ValueError(f"{field} must be {qualifier}")
    return amount


def require_currency(value: object, field: str) -> str:
    """Validate and normalize an ISO-style three-letter currency code."""

    if not isinstance(value, str) or len(value) != 3 or not value.isascii() or not value.isalpha():
        raise ValueError(f"{field} must be a three-letter currency code")
    return value.upper()


def call_billing_event_sink(sink: BillingEventSink, event: BillingEvent) -> BillingEventResult:
    """Dispatch a billing event and raise on unexpected failures."""

    def attempt() -> BillingEventResult:
        result = sink.ingest_billing_event(event)
        if result.error == "claim_busy":
            raise _BillingClaimBusyError
        return result

    try:
        result = retry_bursar_operation(
            attempt,
            retry_options=BursarRetryOptions(
                max_attempts=7,
                base_delay_seconds=0.025,
                max_delay_seconds=0.8,
                max_elapsed_seconds=5.0,
                should_retry=lambda error: isinstance(error, _BillingClaimBusyError),
            ),
        )
    except _BillingClaimBusyError as error:
        raise StoreUnavailableError(
            "Billing event claim remained busy",
            cause=error,
            details={"reason": "claim_busy"},
        ) from error

    if result.error == "claim_failed_retry":
        raise StoreUnavailableError(
            "Billing event claim could not be acquired",
            details={"reason": result.error},
        )
    if not result.handled and result.error not in ("unhandled_event_type", "user_not_found"):
        raise BursarError(
            "Bursar failed to ingest the billing event",
            details={"reason": result.error or "unknown"},
        )
    return result


def parse_subscription_status(raw: object) -> BillingSubscriptionStatus | None:
    """Parse an exact provider status without admitting unknown database values."""
    if not isinstance(raw, str):
        return None
    try:
        return BillingSubscriptionStatus(raw)
    except ValueError:
        return None
